### DISCUSSION TITLE
How to get access token in server?

### DISCUSSION THREAD

**Discussion by hallee9000** [supabase/supabase#6723] [Questions] [answered]
How can I get the user's access token on the server? Or, how can I set the file's owner when uploading files to storage on the server?

**Comment by GaryAustin1** [comment, marked as answer]
Not 100% clear on what you are asking, but most likely you need to pass the users jwt to the server and then use supabase.auth.setAuth(jwt) on the server, then make your storage call.

This thread might have some useful info:
https://github.com/supabase/supabase/discussions/1094

**Reply by hallee9000** [reply]
Hi @GaryAustin1 , thanks for your answer. What I want is to make some requests as a specific user on the server. The key issue is that I can't get jwt from client requests, the only thing I can get is the user id. So I wonder if I can simulate a user by user id on the server.

**Reply by GaryAustin1** [reply]
That seems to be discussed by making a jwt with your secret at this post:
https://github.com/supabase/supabase/discussions/1094#discussioncomment-587587

There is also getUserById in the auth api:
https://github.com/supabase/gotrue-js/blob/8c6373b9fa11982cb5011df0a273498e3196b302/src/GoTrueApi.ts#L675
That will get you a jwt I believe, then you set it with setAuth.

**Reply by hallee9000** [reply]
I tried it but it doesn't return a jwt.
<img width="449" alt="image" src="https://user-images.githubusercontent.com/12868055/167278894-8ce14f9a-9fd8-46e4-af48-0915bff1e9be.png">

**Reply by GaryAustin1** [reply]
I apologize for leading you astray on the getUserById, as I don't do server side stuff myself, but thought there was a post here with someone using that.
So then you are back to the original thread where you know the uuid and you know the secret to generate a jwt and use that with setAuth.
https://supabase.com/docs/learn/auth-deep-dive/auth-deep-dive-jwts
If that does not get you there, then maybe tee up a new question as this one has run out of life....

**Reply by hallee9000** [reply]
Thanks, @GaryAustin1. I think I've understood it. I'll dive into how to implement it.

**Reply by elanzini** [reply]
@leadream did you have any luck getting JWTs? currently facing the same issue.

**Reply by hallee9000** [reply]
@elanzini Finally I figured it out.
1. Get user's info by [`getUserById`](https://github.com/supabase/gotrue-js/blob/8c6373b9fa11982cb5011df0a273498e3196b302/src/GoTrueApi.ts#L675) in server (use server role key to initialize supabase)
2. Generate a jwt token with user's info by [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken)

**Reply by elanzini** [reply]
That’s great to hear. Would you be able to share some code samples? I have little familiarity with supabase.

For example, the id of the user is that the one that I see in the JSON above? So are you sending a request to the server with the user id? Why do you need to generate a JWT on the server side?

My understanding is that the JWT is passed from the client to the server and serves the purpose of authenticating the request is done by a specific user.

**Reply by hallee9000** [reply]
In my case I it's a figma plugin. I can't use supabase sdk in it because it can't access local storage. So I do something like [this](https://github.com/supabase/supabase/discussions/6322). I don't have user's token in plugin so I need to generate one by every user's info.
![image](https://user-images.githubusercontent.com/12868055/192732741-f7e7c1f5-c69a-4ea7-a018-d941b9f911b3.png)

**Comment by a4amado** [comment]
if anyone comes here in the future 

```jsx

type StateType = {
  session: Session | null;
  status: "loading" | "authenticated" | "unauthenticated";
};
export const sessionState = atom<StateType>({
  session: null,
  status: "loading",
});

export default function NextAuthProvider({ children }: PropsWithChildren) {
  const [_, setSession] = useAtom(sessionState);
  const [cookie, setSessionCookie, deleteSessionCookie] = useCookie("session");
  

  useEffect(() => {
    supabaseclient.auth
      .getSession()
      .then(({ data, error }) => {
        setSessionCookie(data.session?.access_token || "")
        setSession({
          session: data.session,
          status: !data.session ? "unauthenticated" : "authenticated",
        });
        supabaseclient.auth.onAuthStateChange((event, session) => {
          setSessionCookie(session?.access_token || "")
          if (event === "SIGNED_OUT") deleteSessionCookie()
          setSession({
            session: event == "SIGNED_OUT" ? null : session,
            status: event == "SIGNED_OUT" ? "unauthenticated" : "authenticated",
          });
        });
      })
      .catch(() => {
        deleteSessionCookie()
        setSession({
          session: null,
          status: "unauthenticated",
        });
      });
  }, []);

  return <>{children}</>;
}

```

state hook 
```jsx
"use client";

import { sessionState } from "./Provider";
import { useAtom } from "jotai";

function useSessionState() {
  const [s] = useAtom(sessionState);
  return s;
}
export default useSessionState;
```

**Reply by frankdagostino** [reply]
**I need to understand this!!!**

ActuallyI'm trying to securely handle access tokens in my server components on Next.js, but I can't trust the session that's passed via a non-HttpOnly cookie from the client side. So, I'm trying to figure out how to keep an HttpOnly cookie always synchronized with the default Supabase cookie to use in my server components."

**Reply by a4amado** [reply]
Supabase is designed for frontend use, similar to Firebase. If you want to use it on the backend, you would need to implement your own authentication on your servers and make your server the mediator between the client and Supabase. However, this approach defeats the main purpose of Supabase, which is to be "serverless." At this point, you are essentially using Supabase as just PostgreSQL.
